srtcore/group.cpp

```diff
@@ -2097,195 +2311,13 @@ int CUDTGroup::recv(char* buf, int len, SRT_MSGCTRL& w_mc)
-
-                read_ready.push_back(gi->ps);
-                HCLOG(ds << "@" << gi->id << "[READ] ");
-            }
-        }
-
-        int read_modes = SRT_EPOLL_IN | SRT_EPOLL_ERR;
-
-        /* Done at the connecting stage so that it won't be missed.
-
-           int connect_modes = SRT_EPOLL_OUT | SRT_EPOLL_ERR;
-           for (vector<SRTSOCKET>::iterator i = connect_pending.begin(); i != connect_pending.end(); ++i)
-           {
-           XXX This is wrong code; this should use the internal function and pass CUDTSocket*
-           epoll_add_usock_INTERNAL(m_RcvEID, i->second, &connect_modes);
-           }
-
-           AND this below additionally for sockets that were so far pending connection,
-           will be now "upgraded" to readable sockets. The epoll adding function for a
-           socket that already is in the eid container will only change the poll flags,
-           but will not re-add it, that is, event reports that are both in old and new
-           flags will survive the operation.
-
-         */
-
-        for (vector<CUDTSocket*>::iterator i = read_ready.begin(); i != read_ready.end(); ++i)
-        {
-            // NOT using the official srt_epoll_add_usock because this will do socket dispatching,
-            // which requires lock on m_GlobControlLock, while this lock cannot be applied without
-            // first unlocking m_GroupLock.
-            CUDT::s_UDTUnited.epoll_add_usock_INTERNAL(m_RcvEID, *i, &read_modes);
-        }
-
-        HLOGC(grlog.Debug, log << "group/recv: " << ds.str() << " --> EPOLL/SWAIT");
-#undef HCLOG
-
-        if (!still_alive)
-        {
-            LOGC(grlog.Error, log << "group/recv: all links broken");
-            throw CUDTException(MJ_CONNECTION, MN_NOCONN, 0);
-        }
-
-        // Here we need to make an additional check.
-        // There might be a possibility that all sockets that
-        // were added to the reader group, are ahead. At least
-        // surely we don't have a situation that any link contains
-        // an ahead-read subsequent packet, because GroupCheckPacketAhead
-        // already handled that case.
-        //
-        // What we can have is that every link has:
-        // - no known seq position yet (is not registered in the position map yet)
-        // - the position equal to the latest delivered sequence
-        // - the ahead position
-
-        // Now the situation is that we don't have any packets
-        // waiting for delivery so we need to wait for any to report one.
-
-        // XXX We support blocking mode only at the moment.
-        // The non-blocking mode would need to simply check the readiness
-        // with only immediate report, and read-readiness would have to
-        // be done in background.
-
-        // Poll on this descriptor until reading is available, indefinitely.
-        CEPoll::fmap_t sready;
+        //bool   still_alive = false;
+        const size_t size = m_Group.size();
```

Comment:
size is now just `size = m_Group.size()`. Previously was incremented inside the loop over the `m_Group`, which is now moved to `CUDTGroup::recv_CollectNonBroken(..)`

Reply:
The intent was to not walk through the whole group list twice. list::size is linear complexity, so this way it passed the loop just once and did two things. Calling m_Group.size() adds one more walk over the group container. You might preserve this by simply collecting it still in the call to `recv_CollectAliveAndBroken` and then return it. OTOH maybe this isn't worth a shot, as the group should normally have at worst up to 4 members...

Reply:
The intent is clear. Just passing all those variables back and force reduces code clarity.
Might be better to count elements inside `GroupContainer` and store list length there. But given there are not many members expected, it might be premature optimization.

Reply:
I agree. So it's ok here.